src/inmanta/execute/dataflow/root_cause.py

```diff
@@ -72,31 +72,51 @@ def root_causes(self) -> Set[AttributeNode]:
             Rules 2 to 4 are implemented as propagation steps by
             _assignment_step, _child_attribute_step and _parent_instance_step respectively.
         """
-        causes: Set[AttributeNode] = set(())
-        for node in self.nodes:
-            others: FrozenSet[AttributeNode] = self.nodes.difference({node})
-            seen: Set[AssignableNode] = set(())
-            to_check: List[AssignableNode] = [node]
-
-            def process_step(step_result: FrozenSet[AssignableNode]) -> None:
-                new: Set[AssignableNode] = set(step_result).difference(seen)
-                to_check.extend(new)
-                seen.update(new)
-
-            is_root_cause: bool = True
-            while to_check:
-                n: AssignableNode = to_check.pop()
-                if n in others and n not in node.equivalence.nodes:
-                    is_root_cause = False
-                    break
-                if n.result_variable is not None and n.result_variable.hasValue:
-                    continue
-                process_step(self._assignment_step(n))
-                process_step(self._parent_instance_step(n))
-                process_step(self._child_attribute_step(n))
-            if is_root_cause:
-                causes.add(node)
-        return causes
+        roots: Set[AttributeNode] = set(())
+        seen: Set[AttributeNode] = set(())
+        to_check: List[AssignableNode] = self.nodes
+
+        def has_root(node: AssignableNode, cycledetect=set()) -> bool:
+            if node in seen:
+                # Already processed, roots are already in roots
+                return True
+
+            if node.value_assignments or (node.result_variable is not None and node.result_variable.hasValue):
+                # Has a value, never has a root cause
+                return False
+
+            if node in cycledetect:
+                # part of a cycle, cycle in itself doesn't provide a root
+                return False
+
+            sub_cycle = cycledetect.union(node.equivalence.nodes)
```

Comment:
This is not correct. Consider the model below:
```
entity A:
    number n
end 

implement A using std::none


x = A()
y = A()
z = A()


x.n = y.n
y.n = x.n
x.n = z.n
```
There's an equivalence between `x.n` and `y.n`. Because you add `x.n` to cycledetect when checking `y.n`, it will not look for a root in `x.n`'s assignments, which it should.
You might need to compile a couple of times to get the faulty output. Because we use sets the order in which the nodes get checked is nondeterministic. I would like to add a test for this but I'm not sure how to deal with the nondeterminism.

Reply:
Tests are run often, so you could let the non-determinism be as it is.

Reply:
I'll add it.